### ISSUE TITLE
Data-Structure ModuleTranslation depends on wasm-data liftetime

### ISSUE THREAD

**Issue by frehberg** [CraneStation/wasmtime#33]
The following function of is creating the ModuleTranslation from raw WASM-data
```rust
impl<'data, 'module> ModuleEnvironment<'data, 'module> {
   pub fn translate(mut self, data: &'data [u8]) -> WasmResult<ModuleTranslation<'data, 'module>> {};
}
```
Due to this, the raw-wasm data must outlive Translation/Compilation. This may cause double memory consumption, keeping the raw-wasm-data and translated/compiled representation in memory.

I would expect that after invocation of function `translate`, the raw WASM-data could be dropped/deleted to reduce memory consumption.

**Comment by sunfishcode** [comment]
I agree, now that this code is becoming more concrete, it's becoming clear that we need to reorganize several of the key data structures.

I think this is the same basic situation as the other issue about `InstanceWorld` too. I'm going to take a look at factoring that better.

**Comment by alexcrichton** [comment]
While technically still applicable to Wasmtime, this is old enough that I'm going to close this out. If this is a concern it's best to have a fresh new issue I believe.
